File: Tmt-server/src/services/task.service.ts
import { TaskRepository }    from '../repositories/task.repository';
import { ProjectRepository } from '../repositories/project.repository';
import { UserRepository }    from '../repositories/user.repository';
import { AppError }          from '../middleware/error.middleware';
import { buildPaginated, parsePagination } from '../utils/pagination';
import { TaskFilters, TaskStatus } from '../types';
import {
  CreateTaskInput,
  UpdateTaskInput,
  AssignTaskInput,
} from '../schemas/task.schema';

export class TaskService {
  constructor(
    private readonly taskRepo:    TaskRepository,
    private readonly projectRepo: ProjectRepository,
    private readonly userRepo:    UserRepository
  ) {}

  async createTask(input: CreateTaskInput, user?: { id: string; role: string }) {
    const project = await this.projectRepo.findById(input.projectId);
    if (!project) throw new AppError(404, 'Project not found');

    if (user?.role === 'EMPLOYEE') {
      const isMember = await this.projectRepo.isMember(input.projectId, user.id);
      if (!isMember) throw new AppError(403, 'You are not a member of this project');
    }

    const finalAssignedTo = input.assignedTo || user?.id;

    if (finalAssignedTo) {
      const assigneeUser = await this.userRepo.findById(finalAssignedTo);
      if (!assigneeUser) throw new AppError(404, 'Assigned user not found');
    }

    return this.taskRepo.create({
      title:       input.title,
      description: input.description,
      projectId:   input.projectId,
      assignedTo:  finalAssignedTo,
      dueDate:     input.dueDate ? new Date(input.dueDate) : undefined,
    });
  }

  async listTasks(query: Record<string, unknown>, user?: { id: string; role: string }) {
    const pagination = parsePagination(query);
    const filters: TaskFilters = {
      projectId:  query.projectId  as string | undefined,
      status:     query.status     as TaskStatus | undefined,
      assignedTo: query.assignedTo as string | undefined,
      title:      query.title      as string | undefined,
      description: query.description as string | undefined,
      dueDate:    query.dueDate    as string | undefined,
    };
    if (user?.role === 'EMPLOYEE') {
      filters.assignedTo = user.id;
    }
    const [tasks, total] = await this.taskRepo.findMany(filters, pagination);
    return buildPaginated(tasks, total, pagination);
  }

  async getTask(id: string, user?: { id: string; role: string }) {
    const task = await this.taskRepo.findByIdWithRelations(id);
    if (!task) throw new AppError(404, 'Task not found');
    if (user?.role === 'EMPLOYEE' && task.assigned_to !== user.id) {
      throw new AppError(403, 'You do not have access to this task');
    }
    return task;
  }

  async updateTask(id: string, input: UpdateTaskInput) {
    const task = await this.taskRepo.findById(id);
    if (!task) throw new AppError(404, 'Task not found');

    if (input.assignedTo) {
      const user = await this.userRepo.findById(input.assignedTo);
      if (!user) throw new AppError(404, 'Assigned user not found');
    }

    return this.taskRepo.update(id, {
      title:       input.title,
      description: input.description,
      status:      input.status as TaskStatus | undefined,
      assignedTo:  input.assignedTo,
      dueDate:     input.dueDate ? new Date(input.dueDate) : (input.dueDate as null | undefined),
    });
  }

  async assignTask(id: string, input: AssignTaskInput) {
    const task = await this.taskRepo.findById(id);
    if (!task) throw new AppError(404, 'Task not found');

    const user = await this.userRepo.findById(input.assignedTo);
    if (!user) throw new AppError(404, 'User not found');

    return this.taskRepo.update(id, { assignedTo: input.assignedTo });
  }

  async deleteTask(id: string) {
    const task = await this.taskRepo.findById(id);
    if (!task) throw new AppError(404, 'Task not found');
    await this.taskRepo.delete(id);
  }
}

File: Tmt-server-go/internal/services/task_service.go
package services

import (
	"context"
	"errors"
	"strings"
	"time"

	appErrors "tmt-server-go/internal/errors"
	"tmt-server-go/internal/pagination"
	"tmt-server-go/internal/repositories"
	"tmt-server-go/internal/types"
)

type TaskService struct {
	tasks    *repositories.TaskRepository
	projects *repositories.ProjectRepository
	users    *repositories.UserRepository
}

func NewTaskService(tasks *repositories.TaskRepository, projects *repositories.ProjectRepository, users *repositories.UserRepository) *TaskService {
	return &TaskService{tasks: tasks, projects: projects, users: users}
}

type CreateTaskInput struct {
	Title       string  `json:"title"`
	Description *string `json:"description"`
	ProjectID   string  `json:"projectId"`
	AssignedTo  *string `json:"assignedTo"`
	DueDate     *string `json:"dueDate"`
}

type UpdateTaskInput struct {
	Title       *string `json:"title"`
	Description *string `json:"description"`
	Status      *string `json:"status"`
	AssignedTo  *string `json:"assignedTo"`
	DueDate     *string `json:"dueDate"`
}

type AssignTaskInput struct {
	AssignedTo string `json:"assignedTo"`
}

func (s *TaskService) CreateTask(ctx context.Context, input CreateTaskInput, user types.JwtUser) (types.Task, error) {
	if strings.TrimSpace(input.Title) == "" {
		return types.Task{}, appErrors.AppError{StatusCode: 422, Message: "title: Title is required"}
	}
	if strings.TrimSpace(input.ProjectID) == "" {
		return types.Task{}, appErrors.AppError{StatusCode: 422, Message: "projectId: Project is required"}
	}

	project, err := s.projects.FindByID(ctx, input.ProjectID)
	if err != nil {
		return types.Task{}, err
	}
	if project == nil {
		return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Project not found"}
	}

	finalAssigned := input.AssignedTo
	if finalAssigned == nil || strings.TrimSpace(*finalAssigned) == "" {
		finalAssigned = &user.ID
	}

	if finalAssigned != nil && strings.TrimSpace(*finalAssigned) != "" {
		assignee, err := s.users.FindById(ctx, *finalAssigned)
		if err != nil {
			return types.Task{}, err
		}
		if assignee == nil {
			return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Assigned user not found"}
		}
	}

	var due *time.Time
	if input.DueDate != nil && strings.TrimSpace(*input.DueDate) != "" {
		parsed, err := parseDate(*input.DueDate)
		if err != nil {
			return types.Task{}, appErrors.AppError{StatusCode: 422, Message: "dueDate: Invalid date format"}
		}
		due = &parsed
	}

	return s.tasks.Create(ctx, struct {
		Title       string
		Description *string
		ProjectID   string
		AssignedTo  *string
		DueDate     *time.Time
	}{
		Title:       input.Title,
		Description: input.Description,
		ProjectID:   input.ProjectID,
		AssignedTo:  finalAssigned,
		DueDate:     due,
	})
}

func (s *TaskService) ListTasks(ctx context.Context, params pagination.CursorParams, user types.JwtUser, filters types.TaskFilters) (types.CursorResult[types.Task], error) {
	return s.tasks.FindMany(ctx, filters, params)
}

func (s *TaskService) GetTask(ctx context.Context, id string, user types.JwtUser) (types.Task, error) {
	task, err := s.tasks.FindByIDWithRelations(ctx, id)
	if err != nil {
		return types.Task{}, err
	}
	if task == nil {
		return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Task not found"}
	}
	return *task, nil
}

func (s *TaskService) UpdateTask(ctx context.Context, id string, input UpdateTaskInput) (types.Task, error) {
	task, err := s.tasks.FindByID(ctx, id)
	if err != nil {
		return types.Task{}, err
	}
	if task == nil {
		return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Task not found"}
	}

	var status *types.TaskStatus
	if input.Status != nil && strings.TrimSpace(*input.Status) != "" {
		sv := types.TaskStatus(strings.TrimSpace(*input.Status))
		if sv != types.TaskTodo && sv != types.TaskInProgress && sv != types.TaskDone {
			return types.Task{}, appErrors.AppError{StatusCode: 422, Message: "status: Invalid status"}
		}
		status = &sv
	}

	if input.AssignedTo != nil && strings.TrimSpace(*input.AssignedTo) != "" {
		user, err := s.users.FindById(ctx, *input.AssignedTo)
		if err != nil {
			return types.Task{}, err
		}
		if user == nil {
			return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Assigned user not found"}
		}
	}

	var due **time.Time
	if input.DueDate != nil {
		if strings.TrimSpace(*input.DueDate) == "" {
			empty := (*time.Time)(nil)
			due = &empty
		} else {
			parsed, err := parseDate(*input.DueDate)
			if err != nil {
				return types.Task{}, appErrors.AppError{StatusCode: 422, Message: "dueDate: Invalid date format"}
			}
			parsedPtr := &parsed
			due = &parsedPtr
		}
	}

	return s.tasks.Update(ctx, id, struct {
		Title       *string
		Description *string
		Status      *types.TaskStatus
		AssignedTo  *string
		DueDate     **time.Time
	}{
		Title:       input.Title,
		Description: input.Description,
		Status:      status,
		AssignedTo:  input.AssignedTo,
		DueDate:     due,
	})
}

func (s *TaskService) AssignTask(ctx context.Context, id string, input AssignTaskInput) (types.Task, error) {
	task, err := s.tasks.FindByID(ctx, id)
	if err != nil {
		return types.Task{}, err
	}
	if task == nil {
		return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "Task not found"}
	}

	assignee, err := s.users.FindById(ctx, input.AssignedTo)
	if err != nil {
		return types.Task{}, err
	}
	if assignee == nil {
		return types.Task{}, appErrors.AppError{StatusCode: 404, Message: "User not found"}
	}

	return s.tasks.Update(ctx, id, struct {
		Title       *string
		Description *string
		Status      *types.TaskStatus
		AssignedTo  *string
		DueDate     **time.Time
	}{
		AssignedTo: &input.AssignedTo,
	})
}

func (s *TaskService) DeleteTask(ctx context.Context, id string) error {
	task, err := s.tasks.FindByID(ctx, id)
	if err != nil {
		return err
	}
	if task == nil {
		return appErrors.AppError{StatusCode: 404, Message: "Task not found"}
	}
	return s.tasks.Delete(ctx, id)
}

func parseDate(s string) (time.Time, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return time.Time{}, errors.New("empty date")
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t, nil
	}
	if t, err := time.Parse("2006-01-02", s); err == nil {
		return t, nil
	}
	return time.Time{}, errors.New("invalid date")
}
